### PR TITLE
Codechange: replace left QSortT() with std::sort()

### PR DESCRIPTION
**Pull request by glx22**
To sort arrays are replaced with std::array or std::vector depending on context.

### CODE REVIEW

**Comment by michicc** [comment]
Using `std::array` where possible is certainly a good idea. JFTR though, you can use `std::sort` on C arrays with `std::begin` and `std::end`.

**Comment by glx22** [comment]
And of course OSX still uses QSortT()

**Comment by glx22** [comment]
Used the easy way for OSX as I can't test it myself.

**Comment by LordAro** [comment]
several of these std::arrays feel like they could be replaced with std::vector, and the _foobar_size globals could be removed as a result

Also, std::begin() & std::end() can be used on C-style arrays

**Review by LordAro — DISMISSED** [review]
As discussed - a few more usages of `std::vector` please :)

File: src/video/allegro_v.cpp [hunk 1]
```diff
@@ -139,52 +139,52 @@ static void GetVideoModes()
...
 	GFX_MODE *modes = mode_list->mode;
 
-	int n = 0;
 	for (int i = 0; modes[i].bpp != 0; i++) {
 		uint w = modes[i].width;
 		uint h = modes[i].height;
 		if (w >= 640 && h >= 480) {
-			int j;
-			for (j = 0; j < n; j++) {
+			uint j;
```
Comment (LordAro) [inline src/video/allegro_v.cpp]: whole block could be replaced with
Dimension might need an `operator==`

```suggestion
			bool has_resolution = std::find(_resolutions.begin(), _resolutions.end(), Dimension(w, h)) == _resolutions.end();
			if (!has_resolution) _resolutions.emplace_back(w, h);
```

File: src/settings_gui.cpp [hunk 2]
```diff
@@ -106,11 +106,11 @@ static inline StringID TownName(int town_name)
  * Get index of the current screen resolution.
  * @return Index of the current screen resolution if it is a known resolution, #_num_resolutions otherwise.
```
Comment (LordAro) [inline src/settings_gui.cpp]: comment out of date

File: src/rail.h [hunk 3]
```diff
@@ -464,14 +464,13 @@ void ResetRailTypes();
...
-extern uint8 _sorted_railtypes_size;
+extern std::vector<RailType> _sorted_railtypes;
 extern RailTypes _railtypes_hidden_mask;
 
 /**
  * Loop header for iterating over railtypes, sorted by sortorder.
  * @param var Railtype.
  */
-#define FOR_ALL_SORTED_RAILTYPES(var) for (uint8 index = 0; index < _sorted_railtypes_size && (var = _sorted_railtypes[index], true) ; index++)
+#define FOR_ALL_SORTED_RAILTYPES(var) for (uint8 index = 0; index < _sorted_railtypes.size() && (var = _sorted_railtypes[index], true) ; index++)
```
Comment (LordAro) [inline src/rail.h:474]: perhaps uint8 -> size_t ?

Possible future PR: convert these to range-based for loops

there are also others like this

File: src/newgrf_config.cpp [hunk 4]
```diff
@@ -740,25 +737,23 @@ void DoScanNewGRFFiles(NewGRFScanCallback *callback)
 		/* Sort the linked list using quicksort.
 		 * For that we first have to make an array, then sort and
 		 * then remake the linked list. */
-		GRFConfig **to_sort = MallocT<GRFConfig*>(num);
+		std::vector<GRFConfig*> to_sort;
```
Comment (LordAro) [inline src/newgrf_config.cpp]: nitpick: `GRFConfig *`

File: src/cargotype.cpp [hunk 5]
```diff
@@ -132,56 +132,54 @@ SpriteID CargoSpec::GetCargoIcon() const
 	return sprite;
 }
 
-const CargoSpec *_sorted_cargo_specs[NUM_CARGO]; ///< Cargo specifications sorted alphabetically by name.
-uint8 _sorted_cargo_specs_size;                  ///< Number of cargo specifications stored at the _sorted_cargo_specs array (including special cargoes).
-uint8 _sorted_standard_cargo_specs_size;         ///< Number of standard cargo specifications stored at the _sorted_cargo_specs array.
+std::vector<const CargoSpec *> _sorted_cargo_specs; ///< Cargo specifications sorted alphabetically by name.
+uint8 _sorted_standard_cargo_specs_size;            ///< Number of standard cargo specifications stored at the _sorted_cargo_specs array.
```
Comment (LordAro) [inline src/cargotype.cpp]: nitpick: at -> in

File: src/video/win32_v.cpp [hunk 6]
```diff
@@ -1085,45 +1085,43 @@ static const Dimension default_resolutions[] = {
...
-				if (++n == lengthof(_resolutions)) break;
 			}
 		}
 	}
 
 	/* We have found no resolutions, show the default list */
-	if (n == 0) {
-		memcpy(_resolutions, default_resolutions, sizeof(default_resolutions));
-		n = lengthof(default_resolutions);
+	if (_resolutions.size() == 0) {
```
Comment (LordAro) [inline src/video/win32_v.cpp]: i'd prefer using `.empty()` here

File: src/video/win32_v.cpp [hunk 7]
```diff
@@ -1085,45 +1085,43 @@ static const Dimension default_resolutions[] = {
...
+		for (uint n = 0; n < lengthof(default_resolutions); n++) _resolutions.push_back(default_resolutions[n]);
 	}
```
Comment (LordAro) [inline src/video/win32_v.cpp:1109]: perhaps `_resolutions = default_resolutions;` ? I think everything would be copied properly...

File: src/rail.h (hunk 3 above)
Comment (glx22) [reply src/rail.h:474]: Range loop is a good idea, but some FOR_ALL uses expect index to be usable.

File: src/video/allegro_v.cpp (hunk 1 above)
Comment (glx22) [reply src/video/allegro_v.cpp]: and a non default constructor too ;)

File: src/highscore.cpp [hunk 8]
```diff
@@ -98,7 +98,7 @@ int8 SaveHighScoreValueNetwork()
 	/* Sort all active companies with the highest score first */
 	FOR_ALL_COMPANIES(c) cl[count++] = c;
 
-	QSortT(cl, count, &HighScoreSorter);
+	std::sort(std::begin(cl), std::begin(cl) + count, HighScoreSorter);
```
Comment (michicc) [inline src/highscore.cpp:100]: `std::end(cl)` instead of `std::begin(cl) + count`?

File: src/industry_gui.cpp [hunk 9]
```diff
@@ -2723,8 +2722,7 @@ struct IndustryCargoesWindow : public Window {
 
 			case WID_IC_IND_DROPDOWN: {
 				DropDownList lst;
-				for (uint i = 0; i < NUM_INDUSTRYTYPES; i++) {
-					IndustryType ind = _sorted_industry_types[i];
+				for (IndustryType ind: _sorted_industry_types) {
```
Comment (michicc) [inline src/industry_gui.cpp]: As far as I can see, current coding style seems to be `for (type a : foo)`.

File: src/industry_gui.cpp [hunk 10]
```diff
@@ -2811,10 +2809,10 @@ const int IndustryCargoesWindow::VERT_TEXT_PADDING = 5; ///< Vertical padding ar
 static void ShowIndustryCargoesWindow(IndustryType id)
 {
 	if (id >= NUM_INDUSTRYTYPES) {
-		for (uint i = 0; i < NUM_INDUSTRYTYPES; i++) {
-			const IndustrySpec *indsp = GetIndustrySpec(_sorted_industry_types[i]);
+		for (IndustryType ind: _sorted_industry_types) {
```
Comment (michicc) [inline src/industry_gui.cpp]: ditto

File: src/highscore.cpp (hunk 8 above)
Comment (glx22) [reply src/highscore.cpp:100]: No because only `count` elements are valid, due to how `FOR_ALL_COMPANIES()` works

**Review by michicc — APPROVED** [review]

**Pull request merged**